frontend/src/metabase/visualizations/echarts/cartesian/model/dataset.ts

```diff
@@ -0,0 +1,129 @@
+ * For unsaved questions without cardId the format is "columnName" or "breakoutValue:columnName" for breakout series.
+ * For saved questions keys include "cardId:" prefix.
+ *
+ * @param {DatasetColumn} column - The series metric column.
+ * @param {number | undefined} cardId - The ID of the card.
+ * @param {RowValue} [breakoutValue] - An optional breakout value when two dimensions columns are selected.
+ * @returns {string} A unique key for the series.
+ */
+export const getDatasetKey = (
+  column: DatasetColumn,
+  cardId?: number,
+  breakoutValue?: RowValue,
+): string => {
+  const cardIdPart = cardId != null ? `${cardId}:` : "";
+  const breakoutPart =
+    typeof breakoutValue !== "undefined" ? `${breakoutValue}:` : "";
+  const columnNamePart = column.name;
+
+  return `${cardIdPart}${breakoutPart}${columnNamePart}`;
+};
+
+/**
+ * Aggregates metric column values in a datum for a given row.
+ * When a breakoutIndex is specified it aggregates metrics per breakout value.
+ *
+ * @param {Record<DataKey, RowValue>} datum - The datum object to aggregate metric values.
+ * @param {DatasetColumn[]} columns - The columns of the raw dataset.
+ * @param {RowValue[]} row - The raw row of values.
+ * @param {number} cardId - The ID of the card.
+ * @param {number} [breakoutIndex] - The breakout column index for charts with two dimension columns selected.
+ */
+const aggregateMetricsForDatum = (
+  datum: Record<DataKey, RowValue>,
+  columns: DatasetColumn[],
+  row: RowValue[],
+  cardId: number,
+  breakoutIndex?: number,
+): void => {
+  columns.forEach((column, columnIndex) => {
+    if (!isMetric(column)) {
+      return;
+    }
+
+    const rowValue = row[columnIndex];
+    const seriesKey = getDatasetKey(column, cardId);
+    datum[seriesKey] = sumMetric(datum[seriesKey], rowValue);
+
+    if (breakoutIndex != null) {
+      const breakoutValue = row[breakoutIndex];
+      const breakoutSeriesKey = getDatasetKey(column, cardId, breakoutValue);
+      datum[breakoutSeriesKey] = sumMetric(datum[breakoutSeriesKey], rowValue);
+    }
+  });
+};
+
+/**
+ * Accepts merged raw cards and raw datasets, groups and joins the metric columns on the dimension column
+ * of each card.
+ *
+ * @param {RawSeries} rawSeries - An array of raw cards merged with raw datasets.
+ * @param {CartesianChartColumns[]} cardsColumns - The columns of each card.
+ * @returns {Record<DataKey, RowValue>[]} The aggregated dataset.
+ */
+export const getJoinedCardsDataset = (
+  rawSeries: RawSeries,
+  cardsColumns: CartesianChartColumns[],
+): Record<DataKey, RowValue>[] => {
+  const groupedData = new Map<RowValue, Record<DataKey, RowValue>>();
+
+  rawSeries.forEach((cardSeries, index) => {
+    const {
+      card,
+      data: { rows, cols },
+    } = cardSeries;
+    const columns = cardsColumns[index];
+
+    const cardId = card.id;
+
+    const dimensionIndex = columns.dimension.index;
+    const dimensionColumn = cols[dimensionIndex];
```

Comment:
Nit: I think this can just be inlined on the next line since we don't reuse it. It's easier to parse the function body when there are less constants being declared, especially for things that aren't reused.